### PR TITLE
Allow creation of multiple control systems from options

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes
Previously, only one control system could be created from options. This allows for an arbitrary number of control systems to be created from options (so long as the control systems themselves exist).

Also added options to the controller which will be needed.

The first couple commits are just small errors I found.

### Upgrade instructions
To add multiple control systems, this is what the yaml file would look like:
```
ControlSystems:
  ControlSystem1:
    Averager:
    Controller:
    TimescaleTuner:
  ControlSystem2:
    Averager:
    Controller:
    TimescaleTuner:
```
where `ControlSystem1` and `ControlSystem2` will be what is returned from the `name()` method of a struct conforming to the ControlSystem protocol.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Helpers/ControlSystem/TestStructs.hpp [hunk 1]
```diff
@@ -13,28 +14,27 @@
 
 namespace control_system::TestHelpers {
 namespace TestStructs_detail {
-struct LabelA;
+struct LabelA {};
```
Comment (wthrowe) [inline tests/Unit/Helpers/ControlSystem/TestStructs.hpp:17]: Wrong commit?  Or maybe unnecessary?

File: src/ControlSystem/Tags.hpp [hunk 2]
```diff
@@ -184,4 +170,66 @@ struct MeasurementTimescales : db::SimpleTag {
...
+  ControlSystem(const ControlSystem& /*rhs*/) = default;
+  ControlSystem& operator=(const ControlSystem& /*rhs*/) = default;
+  ControlSystem(ControlSystem&& /*rhs*/) = default;
+  ControlSystem& operator=(ControlSystem&& /*rhs*/) = default;
+  ~ControlSystem() = default;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) { p | all_inputs_; };
+
+  tuple_of_inputs get_inputs() const { return std::move(all_inputs_); }
```
Comment (wthrowe) [inline src/ControlSystem/Tags.hpp]: You can't move out of a const (because in a const method) member variable.

File: src/ControlSystem/Tags.hpp [hunk 3]
```diff
@@ -184,4 +170,66 @@ struct MeasurementTimescales : db::SimpleTag {
...
+  };
+
+  struct TimescaleTuner {
+    using type = ::TimescaleTuner;
+    static constexpr Options::String help = {
+        "Options for the timescale tuner."};
+  };
+
+  using options = tmpl::list<Averager, Controller, TimescaleTuner>;
+  static constexpr Options::String help = {"Options for a controlsystem."};
```
Comment (wthrowe) [inline src/ControlSystem/Tags.hpp]: "control system"

File: tests/Unit/ControlSystem/Test_Initialization.cpp [hunk 4]
```diff
@@ -44,17 +41,19 @@ struct MockControlComponent {
...
-                                 control_system::Tags::TimescaleTuner,
-                                 control_system::Tags::ControlSystemName,
-                                 control_system::Tags::Controller<2>>;
+  using simple_tags =
+      tmpl::list<control_system::Tags::ControlSystemInputs<MockControlSystem<
+                     LabelA, control_system::TestHelpers::Measurement<LabelA>>>,
+                 control_system::Tags::Averager<2>,
+                 control_system::Tags::TimescaleTuner,
+                 control_system::Tags::ControlSystemName,
+                 control_system::Tags::Controller<2>>;
```
Comment (wthrowe) [inline tests/Unit/ControlSystem/Test_Initialization.cpp:50]: Except for ControlSystemName, everything in this list is added unconditionally by the initialization action.  Is there a reason to duplicate the list here (particularly the initialization-only item)?

File: tests/Unit/ControlSystem/Test_Initialization.cpp [hunk 5]
```diff
@@ -44,17 +41,19 @@ struct MockControlComponent {
...
 
   using phase_dependent_action_list = tmpl::list<Parallel::PhaseActions<
       typename metavariables::Phase, metavariables::Phase::Initialization,
       tmpl::list<ActionTesting::InitializeDataBox<simple_tags>,
                  Initialization::Actions::InitializeControlSystem<
-                     Metavariables, mock_control_sys>,
-                 Initialization::Actions::RemoveOptionsAndTerminatePhase>>>;
+                     Metavariables, mock_control_sys>>>>;
```
Comment (wthrowe) [inline tests/Unit/ControlSystem/Test_Initialization.cpp]: Why this change?

File: tests/Unit/ControlSystem/Test_Initialization.cpp [hunk 6]
```diff
@@ -87,16 +98,31 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.Initialization",
...
+  CHECK(box_averager != averager);
+  CHECK(box_tuner != tuner);
+  CHECK(box_controlsys_name != controlsys_name);
+  // We don't check the controller now because one of its members is
+  // inititalized with signaling_NaN() so comparing now would throw an FPE.
+
+  // Now initialize everything
+  runner.force_next_action_to_be<
+      component, Initialization::Actions::InitializeControlSystem<
+                     MockMetavars, mock_control_sys>>(0);
```
Comment (wthrowe) [inline tests/Unit/ControlSystem/Test_Initialization.cpp]: This is already the next action, isn't it?  Is there something I'm missing going on here?

File: src/ControlSystem/Controller.hpp [hunk 7]
```diff
@@ -38,17 +40,62 @@ class er;
...
-  void pup(PUP::er& /*p*/) {}
-
   DataVector operator()(
       const DataVector& timescales,
       const std::array<DataVector, DerivOrder + 1>& q_and_derivs,
       double q_time_offset, double deriv_time_offset) const;
+
+  // This will be necessary for when a function exists that actually updates a
+  // control system. Adding it in now because it uses update_fraction_ which is
+  // read in from options.
```
Comment (wthrowe) [inline src/ControlSystem/Controller.hpp]: Remove comment.  This function is used in this PR.

File: src/ControlSystem/Actions/Initialization.hpp [hunk 8]
```diff
@@ -58,6 +69,32 @@ struct InitializeControlSystem {
...
+        std::move(tuples::get<control_system::Tags::Averager<deriv_order>>(
+            control_sys_inputs)),
+        std::move(tuples::get<control_system::Tags::Controller<deriv_order>>(
+            control_sys_inputs)),
+        std::move(tuples::get<control_system::Tags::TimescaleTuner>(
+            control_sys_inputs)));
+
+    // Set the initial time between updates using the initial timescale
+    const auto& tuner = db::get<control_system::Tags::TimescaleTuner>(box);
+    const double current_min_timescale = min(tuner.current_timescale());
```
Comment (wthrowe) [inline src/ControlSystem/Actions/Initialization.hpp:78]: Maybe not for this PR, but we will have to guarantee somehow that there are no control system updates attempted before the last time used during integrator self-start.  The only idea I've had is tying the first update to the slab size, but that might be confusing because that's probably not an expected effect of changing the slab size.

File: src/ControlSystem/Controller.hpp [hunk 9]
```diff
@@ -38,17 +40,62 @@ class er;
...
+  struct UpdateFraction {
+    using type = double;
+    static constexpr Options::String help = {
+        "Fraction of damping timescale used to determine how often to update "
+        "functions of time."};
+  };
+
+  using options = tmpl::list<UpdateFraction>;
+  static constexpr Options::String help{
+      "Controller. Computes control signal used to reset highest derivative of "
```
Comment (wthrowe) [inline src/ControlSystem/Controller.hpp]: Remove "Controller."  The class being created is included in the error message.

File: src/ControlSystem/Tags.hpp [hunk 10]
```diff
@@ -184,4 +170,66 @@ struct MeasurementTimescales : db::SimpleTag {
   }
 };
 }  // namespace Tags
+
+/// \ingroup ControlSystemGroup
+/// Place to hold all options for a single control system
+namespace OptionHolder {
```
Comment (wthrowe) [inline src/ControlSystem/Tags.hpp]: Is there anything else that will go in this namespace?  If not, let's use it as the class name instead.

File: src/ControlSystem/Tags.hpp [hunk 11]
```diff
@@ -184,4 +170,66 @@ struct MeasurementTimescales : db::SimpleTag {
...
+        "Options for the timescale tuner."};
+  };
+
+  using options = tmpl::list<Averager, Controller, TimescaleTuner>;
+  static constexpr Options::String help = {"Options for a controlsystem."};
+
+  using db_tags = tmpl::list<control_system::Tags::Averager<DerivOrder>,
+                             control_system::Tags::Controller<DerivOrder>,
+                             control_system::Tags::TimescaleTuner>;
+  using tuple_of_inputs = tuples::tagged_tuple_from_typelist<db_tags>;
```
Comment (wthrowe) [inline src/ControlSystem/Tags.hpp]: `TupleOfInputs` because its a "real" class as opposed to a metavariable.

File: src/ControlSystem/Tags.hpp [hunk 12]
```diff
@@ -184,4 +170,66 @@ struct MeasurementTimescales : db::SimpleTag {
...
+/// Place to hold all options for a single control system
+namespace OptionHolder {
+/// \ingroup ControlSystemGroup
+/// Holds all options for a single control system
+///
+/// This struct collects all the options for a given control system during
+/// option parsing. Then during initialization, the options can be retreived as
+/// a TaggedTuple and assigned to their corresponding DataBox tags.
+template <size_t DerivOrder>
+struct ControlSystem {
```
Comment (wthrowe) [inline src/ControlSystem/Tags.hpp]: This class would be simpler as a "C-style" struct with public member variables.  It wouldn't need any member functions except `pup`.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Helpers/ControlSystem/TestStructs.hpp (hunk 1 above)
Comment (knelli2) [reply tests/Unit/Helpers/ControlSystem/TestStructs.hpp:17]: Oh yeah probably wrong commit. I use this label in the "Add tags for multiple control systems" commit as the label for the test control system. The compiler complained about it being just a forward declaration and not a definition. I must have just lumped it in with the changes to this file. I'll change it.

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 4 above)
Comment (knelli2) [reply tests/Unit/ControlSystem/Test_Initialization.cpp:50]: So this may be because I'm not super familiar with how to use the Action Testing framework, but I originally tried just
```
using simple_tags = tmpl::list<control_system::Tags::ControlSystemName>;
```
but I got a ton of compiler errors that essentially said it couldn't find any of the other tags in the DataBox. And I wasn't sure why. I thought it was because I just had to add everything manually, so that's what I did. It appeared like the `InitializeDataBox` action didn't find the initialization tags and simple tags from my `InitializeControlSystem` action. Am I initializing things incorrectly?

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 5 above)
Comment (knelli2) [reply tests/Unit/ControlSystem/Test_Initialization.cpp]: I realized I never used this action in the test so I just removed it. Do I need to test it?

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 6 above)
Comment (knelli2) [reply tests/Unit/ControlSystem/Test_Initialization.cpp]: This may have something to do with the previous comment about initialization, I'm not sure. But I tried `ActionTesting::next_action` like you are suggesting and that gave an error. Then I tried to just see how many actions were left with `ActionTesting::number_of_queued_simple_actions` and it was 0. Even though I hadn't ran any actions. I was very confused by this.
So this was the only way I could get the action to run.

File: src/ControlSystem/Tags.hpp (hunk 10 above)
Comment (knelli2) [reply src/ControlSystem/Tags.hpp]: No nothing else is in this namespace. I'll use it as the class name instead.

File: src/ControlSystem/Tags.hpp (hunk 12 above)
Comment (knelli2) [reply src/ControlSystem/Tags.hpp]: Ok I'll just expose the member variables. Then I don't have to bother with the `TaggedTuple`s and such.

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 4 above)
Comment (wthrowe) [reply tests/Unit/ControlSystem/Test_Initialization.cpp:50]: Never mind, you're right.  I misread the code.

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 5 above)
Comment (wthrowe) [reply tests/Unit/ControlSystem/Test_Initialization.cpp]: OK, removing it is fine.

File: tests/Unit/ControlSystem/Test_Controller.cpp [hunk 13]
```diff
@@ -13,9 +13,11 @@
 #include "DataStructures/DataVector.hpp"
 #include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
 #include "Domain/FunctionsOfTime/PiecewisePolynomial.hpp"
+#include "Framework/TestHelpers.hpp"
 #include "Utilities/ConstantExpressions.hpp"
 
-SPECTRE_TEST_CASE("Unit.ControlSystem.Controller", "[ControlSystem][Unit]") {
+void test_controller() {
```
Comment (wthrowe) [inline tests/Unit/ControlSystem/Test_Controller.cpp:20]: Put in anonymous namespace.  Sorry, missed this before.

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 6 above)
Comment (wthrowe) [reply tests/Unit/ControlSystem/Test_Initialization.cpp]: It's not a simple action, it's an iterable action, so that part is expected.

Testing, it seems to work without this.  What problem are you seeing?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/ControlSystem/Test_Initialization.cpp (hunk 6 above)
Comment (knelli2) [reply tests/Unit/ControlSystem/Test_Initialization.cpp]: I think I was just using the ActionTesting framework incorrectly. I got it to work now.

**Comment by knelli2** [comment]
@wthrowe added fixups. The force push of the original commits came from me moving the `struct LabelA{}` to the proper commit. Nothing else was changed there.

File: src/ControlSystem/Tags.hpp [hunk 14]
```diff
@@ -199,37 +194,32 @@ struct ControlSystem {
...
+    p | tuner_;
+  };
 
- private:
-  tuple_of_inputs all_inputs_{};
+  // These members are specifically made pubic for easy access during
+  // initialization
+  ::Averager<DerivOrder> averager_{};
+  ::Controller<DerivOrder> controller_{};
+  ::TimescaleTuner tuner_{};
```
Comment (wthrowe) [inline src/ControlSystem/Tags.hpp]: No trailing underscores on public members.

File: tests/Unit/ControlSystem/Test_Controller.cpp (hunk 13 above)
Comment (wthrowe) [reply tests/Unit/ControlSystem/Test_Controller.cpp:20]: Ping.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Squash, including these two requests.

**Review by wthrowe — DISMISSED** [review]

**Comment by knelli2** [comment]
Forgot to update the docs for `OptionHolder` when I changed the implementation. No other changes.

File: src/ControlSystem/Actions/Initialization.hpp [hunk 15]
```diff
@@ -41,13 +47,18 @@ struct InitializeControlSystem {
   // at the moment, because this is the DerivOrder of most functions of time in
   // the domain creators. If we want to choose the DerivOrder at runtime (i.e.
   // in an input file), that capability will need to be added later.
-  using initialization_tags = tmpl::list<control_system::Tags::Averager<2>,
-                                         control_system::Tags::TimescaleTuner>;
+  static constexpr size_t deriv_order = ControlSystem::deriv_order;
```
Comment (nilsvu) [inline src/ControlSystem/Actions/Initialization.hpp:46]: Update the comment above

File: src/ControlSystem/Tags.hpp [hunk 16]
```diff
@@ -184,4 +168,59 @@ struct MeasurementTimescales : db::SimpleTag {
...
+/// Holds all options for a single control system
+///
+/// This struct collects all the options for a given control system during
+/// option parsing. Then during initialization, the options can be retrieved via
+/// their public member names and assigned to their corresponding DataBox tags.
+template <size_t DerivOrder>
+struct OptionHolder {
+  struct Averager {
+    using type = ::Averager<DerivOrder>;
+    static constexpr Options::String help = {"Options for the averager."};
```
Comment (nilsvu) [inline src/ControlSystem/Tags.hpp]: Please describe these three options a bit in their help string. A help string that just repeats the name of the options is not very helpful ;) You don't even have to say "options for ..." since everything in the input file are options. Just give the user an idea what the averager etc is, and what they should keep in mind for setting its options (if any).

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nilsvu** [comment]
Oh and I think you have to squash the last two commits together, since they don't compile individually. You can also squash my comments above.

File: src/ControlSystem/Actions/Initialization.hpp (hunk 15 above)
Comment (knelli2) [reply src/ControlSystem/Actions/Initialization.hpp:46]: Removed it because it no longer applied here :)

**Comment by knelli2** [comment]
Squashed in @nilsleiffischer's changes and rebased to fix conflict with #3589.

**Comment by wthrowe** [comment]
Needs yet another rebase, but otherwise looks fine.

**Review by wthrowe — APPROVED** [review]

File: src/ControlSystem/Tags.hpp [hunk 17]
```diff
@@ -184,4 +168,64 @@ struct MeasurementTimescales : db::SimpleTag {
...
+  ~OptionHolder() = default;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) {
+    p | averager;
+    p | controller;
+    p | tuner;
+  };
+
+  // These members are specifically made pubic for easy access during
```
Comment (nilsvu) [inline src/ControlSystem/Tags.hpp:225]: typo: `public`. You can fix this in another PR.

**Review by nilsvu — APPROVED** [review]
